### ISSUE TITLE
Improved build documentation when building dependencies

### ISSUE THREAD

**Issue by smartalecH** [JuliaLang/julia#47631]
I've been working on building Julia (release 1.8.2) on some rather complicated internal infrastructure. Throughout my journey (discussed a bit in this [discourse thread](https://discourse.julialang.org/t/installation-clobbers-ldflags/90139)), I've realized there are some important details missing in the build documentation that would probably help others. For example...

1. `patchelf` effectively clobbers the user-specified `LDFLAGS`, and instead uses `CXXLDFLAGS`:

https://github.com/JuliaLang/julia/blob/d12ac369a4be6ae7833fb95c644cea4531764e31/deps/patchelf.mk#L23

2. Some libraries, like `gmp`, effectively _ignore_ `LDFLAGS` altogether, and instead expect the user to pass important parameters, like runtime paths, to the `CFLAGS` and `CXXFLAGS` (as discussed [here](https://gmplib.org/list-archives/gmp-bugs/2012-April/002608.html)).
3. Perhaps the most difficult issue to navigate is the infamous `libstdc++` loading problem (discussed [here](https://github.com/JuliaLang/julia/issues/34276#issuecomment-1258011553)). While there was a recent fix intended to help mitigate this issue (#46976) it's not clear to what extent it does. Furthermore, it appears that PR wasn't percolated to the recent 1.8.3 release, or even the 1.9-alpha release. While a _minimum_ `gcc` version is documented, there is no _maximum_ version documented for any of the releases.

I'm still trying to work through issue 3 myself. When I build LLVM, it uses my compiler's `libstdc++`. When the build script moves to building Julia (and loads some LLVM libraries), the linker references Julia's `libstdc++` rather than the version LLVM was built with. I'm assuming this is only a problem because my compiler's version is newer (a recently patched version of gcc11.2) but I'm still working through this...

I'm happy to add a PR that describes these details more thoroughly, but I want to make sure I'm not missing something obvious first.

Thanks!

**Comment by vtjnash** [comment]
The minimum gcc is better autodetected on newer versions of julia already. But we are not going to backport it to all versions. A more reliable solution can be to set `USE_BINARYBUILDER=0` when you know that binaries are not available from there. And while we try to do some regularization of all the possible CFLAGS, eventually that becomes too big of a problem and you have to do your own digging to figure out what each specific library expects.

The `patchelf` binary may be special, in being compiled for the host architecture, instead of the target? Apparently it was just thought to be a C++ binary though, so was expected at some point to use the alternate flags:
https://github.com/JuliaLang/julia/commit/44e36da185ef7b54c834d387dbb05be811f33531

**Comment by smartalecH** [comment]
Thanks, @vtjnash for your quick reply!

> The minimum gcc is better autodetected on newer versions of julia already. But we are not going to backport it to all versions.

Sorry, are you talking about _maximum_ gcc version? That's the issue I was referring to. 

And I understand if you don't want to backport to all versions. But 1.8.3 was just released last week, right?

> A more reliable solution can be to set USE_BINARYBUILDER=0 when you know that binaries are not available from there.

Yes, all the issues I'm describing above are when `USE_BINARYBUILDER=0`... perhaps I'm misunderstanding something?

> And while we try to do some regularization of all the possible CFLAGS,

I actually don't think you need to regularize all the flags (unless they start clashing). Rather, just adding a few hints about this in the docs would be useful. Or even something like "be warned, each dependency may require careful consideration when setting the appropriate compiler flags".

It's a little odd that the docs spend significant time describing how to add the binary to the `PATH` ( something straightforward) but don't describe any of the caveats with respect to building the dependencies themselves.

**Comment by vtjnash** [comment]
yeah, I was thinking of the minimum for the maximum, so really just the maximum.

I tagged that PR for backport, so we will see how that goes (I was mixing this up with https://github.com/JuliaLang/julia/issues/47629, which I was attempting to answer at the same time–about v1.3 rather than v1.8.3)

That would probably be a good note for the `docs/src/devdocs/build/build.md` file! Would you be willing to write something there, especially given your experience now?

**Comment by smartalecH** [comment]
> That would probably be a good note for the docs/src/devdocs/build/build.md file! Would you be willing to write something there, especially given your experience now?

Sure, I'd be happy to 😀
